drivers/xen/gnttab.c

```diff
@@ -0,0 +1,336 @@
+		unsigned long mfn)
+{
+	gnttab.table[gref].frame = mfn;
+	gnttab.table[gref].domid = domid;
+
+	compiler_barrier();
+}
+
+static void gnttab_grant_permit_access(grant_ref_t gref, domid_t domid,
+		unsigned long mfn, int readonly)
+{
+	gnttab_grant_init(gref, domid, mfn);
+	readonly *= GTF_readonly;
+	gnttab.table[gref].flags = GTF_permit_access | readonly;
+}
+
+grant_ref_t gnttab_grant_access(domid_t domid, unsigned long mfn,
+		int readonly)
+{
+	grant_ref_t gref = get_free_entry();
+
+	gnttab_grant_permit_access(gref, domid, mfn, readonly);
+
+	return gref;
+}
+
+grant_ref_t gnttab_grant_transfer(domid_t domid, unsigned long mfn)
+{
+	grant_ref_t gref = get_free_entry();
+
+	gnttab_grant_init(gref, domid, mfn);
+	gnttab.table[gref].flags = GTF_accept_transfer;
+
+	return gref;
+}
+
+/* Reset flags to zero in order to stop using the grant */
+static int gnttab_reset_flags(grant_ref_t gref)
+{
+	uint16_t flags, nflags;
+	uint16_t *pflags;
+
+	pflags = &gnttab.table[gref].flags;
+	nflags = *pflags;
+
+	do {
+		flags = nflags;
+		if (flags & (GTF_reading | GTF_writing)) {
+			LOG_WRN("gref = %u still in use! (0x%x)\n",
+				gref, flags);
+			return 0;
+		}
+		nflags = synch_cmpxchg(pflags, flags, 0);
+	} while (nflags != flags);
+
+	return 1;
+}
+
+int gnttab_update_grant(grant_ref_t gref, domid_t domid,
+		unsigned long mfn, int readonly)
+{
+	int rc;
+
+	__ASSERT((gref >= GNTTAB_NR_RESERVED_ENTRIES &&
+		gref < NR_GRANT_ENTRIES), "Invalid gref = %d", gref);
+
+	rc = gnttab_reset_flags(gref);
+	if (!rc)
+		return rc;
+
+	gnttab_grant_permit_access(gref, domid, mfn, readonly);
+
+	return 1;
+}
+
+int gnttab_end_access(grant_ref_t gref)
+{
+	int rc;
+
+	__ASSERT((gref >= GNTTAB_NR_RESERVED_ENTRIES &&
+		gref < NR_GRANT_ENTRIES), "Invalid gref = %d", gref);
+
+	rc = gnttab_reset_flags(gref);
+	if (!rc) {
+		return rc;
+	}
+
+	put_free_entry(gref);
+
+	return 1;
+}
+
+unsigned long gnttab_end_transfer(grant_ref_t gref)
+{
+	unsigned long frame;
+	uint16_t flags;
+	uint16_t *pflags;
+
+	__ASSERT((gref >= GNTTAB_NR_RESERVED_ENTRIES &&
+		gref < NR_GRANT_ENTRIES), "Invalid gref = %d", gref);
+
+	pflags = &gnttab.table[gref].flags;
+	flags = *pflags;
+	while (!(flags & GTF_transfer_committed)) {
+		if (synch_cmpxchg(pflags, flags, 0) == flags) {
+			LOG_WRN("Release unused transfer grant.\n");
+			put_free_entry(gref);
+			return 0;
+		}
+		flags = *pflags;
+	}
+
+	/* If a transfer is in progress then wait until it is completed. */
+	while (!(flags & GTF_transfer_completed)) {
+		flags = *pflags;
+	}
+
+	/* Read the frame number /after/ reading completion status. */
+	compiler_barrier();
+	frame = gnttab.table[gref].frame;
+
+	put_free_entry(gref);
+
+	return frame;
+}
+
+grant_ref_t gnttab_alloc_and_grant(void **map)
+{
+	void *page;
+	unsigned long mfn;
+	grant_ref_t gref;
+
+	__ASSERT_NO_MSG(map != NULL);
+
+	page = k_aligned_alloc(XEN_PAGE_SIZE, XEN_PAGE_SIZE);
+	if (page == NULL) {
+		return -ENOMEM;
```

Comment:
grant_ref_t is a uint32_t. So this will result to a large. Shouldn't the function return int32_t instead?

Reply:
Yes, you are right, fixed with int32_t.